src/main/java/org/reactome/release/common/dataretrieval/cosmic/COSMICFileRetriever.java

```diff
@@ -0,0 +1,122 @@
+	{
+		super(retrieverName);
+	}
+	
+	public COSMICFileRetriever()
+	{
+		super();
+	}
+	
+	private boolean retrieveAndSetCOSMICDownloadURL() throws UnsupportedEncodingException
+	{
+		boolean gotDownloadURLOK = false;
+		// The NEW process to download from COSMIC requires a few steps.
+		// 1) Generate a base64-encoded string of the username and password.
+		// 2) Send this encoded string in the "Authorization: Basic" header to the URL of the file you want.
+		// 3) Parse the response - extract the "url" JSON attribute and then download THAT URL to get the REAL file.
+		
+		// Encoded string.
+		String encodedUsernamePassword = Base64.getEncoder().encodeToString((this.userName + ":" + this.password).getBytes("UTF-8"));
+		
+		HttpGet get = new HttpGet(this.uri);
+		//Need to multiply by 1000 because timeouts are in milliseconds.
+		int delayInMilliseconds = 1000 * (int)this.timeout.getSeconds();
+		RequestConfig config = RequestConfig.copy(RequestConfig.DEFAULT)
+											.setConnectTimeout(delayInMilliseconds)
+											.setSocketTimeout(delayInMilliseconds)
+											.setConnectionRequestTimeout(delayInMilliseconds).build();
+		get.setConfig(config);
+		get.setHeader("Authorization", "Basic "+encodedUsernamePassword);
+		String downloadURL = null;
+		try( CloseableHttpClient client = HttpClients.createDefault();
+				CloseableHttpResponse response = client.execute(get) )
+		{
+				int statusCode = response.getStatusLine().getStatusCode();
+				String responseString = EntityUtils.toString(response.getEntity());
+				// If status code was not 200, we should print something so that the users know that an unexpected response was received.
+				switch (statusCode)
+				{
+					case HttpStatus.SC_OK:
+					// Now we need to turn parse the JSON in responseString and extract the URL to download from.
+					JsonReader reader = Json.createReader(new StringReader(responseString));
+					JsonObject responseObject = reader.readObject();
+					downloadURL = responseObject.get("url").toString().replaceAll("\"", "");
+					// Update this object's downloadURL to be the one that came back from the request
+					this.setDataURL(new URI(downloadURL));
+					logger.info("COSMIC download URL has been set.");
+					gotDownloadURLOK = true;
+					// Call downloadData of FileRetriever to perform a "normal" download, now that the special URL has been set.
+					break;
+					
+					default:
+					logger.error("Non-200 status code: {} Response String is: {}", statusCode, responseString);
+					gotDownloadURLOK = false;
+					break;
+				}
+		}
+		catch (IOException e)
+		{
+			logger.error("IOException was caught, probably caused by some network communication issue. Message: {}", e.getMessage());
+			e.printStackTrace();
+		}
+		catch (URISyntaxException e)
+		{
+			logger.error("The URL from COSMIC might be malformed. URL is: \"{}\", Error message is: {}", downloadURL, e.getMessage() );
+			e.printStackTrace();
+		}
```

Comment:
I think one of the recent releases had COSMIC require the owner of the account to 'verify' their account on the actual cosmic website. I couldn't download the file because authentication wasn't working programmatically, and so after getting the credentials I logged in on the website... at which point the programmatic file downloads worked. Any way to guard against that type of situation? I can't recreate it at the moment since it was just something that they seemed to do after a set amount of time (I'm guessing like a year)

Reply:
I didn't even know that such a situation could occur! It seems like an issue very specific to COSMIC. If I know what HTTP response code and headers they sent back in this situation, and could easily test any code I wrote for it, it _might_ be possible, but it seems like a pretty rare occurrence to need to program for.

Reply:
Yea, I wouldn't worry about doing anything else - its an easy fix and I don't know how we'd recreate it anyways